### PR TITLE
[Feat] TensorDictBase.masked_select_ inplace

### PR DESCRIPTION
**Pull request by duanlian-lab**
## Description

See the feature request https://github.com/pytorch/rl/issues/1178

## Types of changes

- [x] New feature (non-breaking change which adds core functionality) 
- [x] Documentation (update in the documentation)
- [x] Example inside the codes

### CODE REVIEW

File: tensordict/tensordict.py [hunk 1]
```diff
@@ -2348,6 +2348,24 @@ def masked_fill(self, mask: Tensor, value: float | bool) -> TensorDictBase:
...
+        Examples:
+            >>> td = TensorDict(source={'a': torch.zeros(3, 4)},
+            ...    batch_size=[3])
+            >>> mask = torch.tensor([True, False, False])
+            >>> td.masked_select_(mask)
+            >>> td.get("a")
+            tensor([[0., 0., 0., 0.]])
+
+        """
+        raise NotImplementedError
```
Comment (vmoens) [inline tensordict/tensordict.py:2367]: Since we implement it for one class only, let's raise a more proper error message, something like
```suggestion
        raise NotImplementedError(f"class {type(self)} does not implement masked_select_. Convert to a `TensorDict` instance first. If this is needed, open an issue on tensordict's repository.")
```

**Review by vmoens — COMMENTED** [review]
Thanks for much for this!
Can you have a look at these few comments?

**Comment by vmoens** [comment]
We should test that feature too!

File: tensordict/tensordict.py (hunk 1 above)
Comment (duanlian-lab) [reply tensordict/tensordict.py]: Thanks

File: tensordict/tensordict.py [hunk 2]
```diff
@@ -4006,6 +4024,55 @@ def to(tensor):
...
+
+        Examples:
+            >>> td = TensorDict(source={'a': torch.zeros(3, 4)},
+            ...    batch_size=[3])
+            >>> mask = torch.tensor([True, False, False])
+            >>> td.masked_select_(mask)
+            >>> td_mask.get("a")
+            tensor([[0., 0., 0., 0.]])
+
+        """
```
[the remark this answers is not in the document]
Comment (duanlian-lab) [reply tensordict/tensordict.py]: Of course :),  Dropped it in a new commit.

File: tensordict/tensordict.py [hunk 3]
```diff
@@ -4006,6 +4024,55 @@ def to(tensor):
...
+    #     Examples:
+    #         >>> td = TensorDict(source={'a': torch.zeros(3, 4)},
+    #         ...    batch_size=[3])
+    #         >>> mask = torch.tensor([True, False, False])
+    #         >>> td_mask = td.masked_select(mask)
+    #         >>> td_mask.get("a")
+    #         tensor([[0., 0., 0., 0.]])
+
+    #     """
+    #     return self.clone().masked_select_(mask)
```
[the remark this answers is not in the document]
Comment (duanlian-lab) [reply tensordict/tensordict.py]: Well, first why I commit these codes?  I wanna provide a semantic-similar implementation for `.masked_select()` and `.masked_select_()`, but the new implementation of `.masked_select()` inside `TensorDict` will break the compatibility sometimes with `.masked_select()` inside `TensorDictBase` ( no more process on the input mask inside the new implementation). So I commented it out.

Then, what to do? We may:
1. Remove it or ignore it. Keep the `.mask_select()` implementation original.
2. Comment in to accept a new implementation for `TensorDict`.

**Comment by duanlian-lab** [comment]
Testing codes added

File: tensordict/tensordict.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vmoens) [reply tensordict/tensordict.py]: Sorry I don't really understand the issue with the current mask_select, can you elaborate further?
